lib/semantics/check-allocate.cc

```diff
@@ -0,0 +1,411 @@
+                      },
+                  },
+                  statOrErr.u);
+            },
+            [&](const parser::AllocOpt::Source &source) {
+              analyzer.Analyze(source.v);
+              if (info.gotSrc) {  // C943
+                context.Say(
+                    "SOURCE may not be duplicated in a ALLOCATE statement"_err_en_US);
+                stopCheckingAllocate = true;
+              }
+              if (info.gotMold || info.gotTypeSpec) {  // C944
+                context.Say(
+                    "At most one of source-expr and type-spec may appear in a ALLOCATE statement"_err_en_US);
+                stopCheckingAllocate = true;
+              }
+              parserSourceExpr = &source.v.value();
+              info.gotSrc = true;
+            },
+            [&](const parser::AllocOpt::Mold &mold) {
+              analyzer.Analyze(mold.v);
+              if (info.gotMold) {  // C943
+                context.Say(
+                    "MOLD may not be duplicated in a ALLOCATE statement"_err_en_US);
+                stopCheckingAllocate = true;
+              }
+              if (info.gotSrc || info.gotTypeSpec) {  // C944
+                context.Say(
+                    "At most one of source-expr and type-spec may appear in a ALLOCATE statement"_err_en_US);
+                stopCheckingAllocate = true;
+              }
+              parserSourceExpr = &mold.v.value();
+              info.gotMold = true;
+            },
+        },
+        allocOpt.u);
+  }
+
+  if (stopCheckingAllocate) {
+    return std::nullopt;
+  }
+
+  if (info.gotSrc || info.gotMold) {
+    CHECK(parserSourceExpr);
+    CHECK(parserSourceExpr->typedExpr);  // TODO: Can we reach this spot without
+                                         // a valid expression?
+    info.sourceExprType = parserSourceExpr->typedExpr->v.GetType();
+    if (!info.sourceExprType.has_value()) {
+      context.Say(parserSourceExpr->source,
+          "Source expression in ALLOCATE must be a valid expression"_err_en_US);
+      return std::nullopt;
+    }
+  }
+  return info;
+}
+
+bool AllocationChecker::RunChecks(SemanticsContext &context) {
+  if (!IsVariableName(*name_.symbol)) {  // C932 pre-requisite
+    context.Say(name_.source,
+        "name in ALLOCATE statement must be a variable name"_err_en_US);
+    return false;
+  }
+  if (!IsAllocatableOrPointer(*name_.symbol)) {  // C932
+    context.Say(name_.source,
+        "%s in ALLOCATE statement must have the ALLOCATABLE or POINTER attribute"_err_en_US,
+        (isSubobject_ ? "component" : "name"));
+    return false;
+  }
+  bool gotSourceExprOrTypeSpec{allocateInfo_.gotMold ||
+      allocateInfo_.gotTypeSpec || allocateInfo_.gotSrc};
+  if (hasDefferedTypeParameter_ && !gotSourceExprOrTypeSpec) {
+    // C933
+    context.Say(name_.source,
+        "Either type-spec or source-expr shall appear in ALLOCATE when allocatable object has a deferred type parameters"_err_en_US);
```

Comment:
I think "must" is better than "shall" in all of these error messages. They use "shall" in the Fortran standard, but as an English word it is ambiguous. It can mean "must" or "should" or "will", so we should avoid it.

Reply:
Sounds easier and more consistent with other error messages, thanks. Applied.